### PR TITLE
Pokémon Red and Blue: Version 4 update

### PR DESCRIPTION
**Pull request by Alchav**
## What is this fixing or adding?
Adds a large number of new options, including:

- Door Shuffle
- Sphere-based level scaling
- Key Item and Pokedex requirement options to reach the Elite Four
- Split Card Key option
- Dexsanity option can be set to a percentage of Pokémon that will be checks
- Stonesanity: remove the stones from the Celadon Department Store and shuffle them into the item pool, replacing 4 of the 5 Moon Stone items
- Sleep Trap items option
- Randomize Move Types option
- Town Map Fly Location option, to unlock a flight location when finding/receiving the Town Map

Many enhancements have been made, including:
- Game allows you to continue your save file _from Pallet Town_ as a way to save warp back to the beginning of the game. The one-way drop from Diglett's Cave to north Route 2 that had been added to the randomizer has been removed.
- Client auto-hints some locations when you are able to see the item before you can obtain it (but would only show AP Item if it is for another player), including Bike Shop, Oak's Aides, Celadon Prize Corner, and the unchosen Fossil location.

Various bugs have been fixed, including:
- Route 13 wild Pokémon not correctly logically requiring Cut
- Vanilla tm/hm compatibility options giving compatibility for many TMs/HMs erroneously 
- If an item that exists in multiple quantities in the item pool is chosen for one of the locations that are pre-filled with local items, it will continue placing that same item in the remaining locations as long as more of that item exist
- `start_with` option for `randomize_pokedex` still shuffling a Pokédex into the item pool
- The obedience threshold levels being incorrect with 0-2 badges, with Pokémon up to level 30 obeying with 0-1 badges and up to 10 with 2 badges
- Receiving a DeathLink trigger in the Safari Zone causing issues. Now, you will have your steps remaining set to 0 instead of blacking out when you're in the Safari Zone.

Many location names have been changed, as location names are automatically prepended using the Region name and a large number of areas have been split into new regions as part of the overhaul to add Door Shuffle.

Adds an optional parameter to `sweep_for_events` for a specific player to sweep for. This is used in door shuffle.

## How was this tested?
Many solo tests have been done, as well as one sync test and one ongoing async test game.

### CODE REVIEW

**Comment by ThePhar** [comment]
> Adds an optional parameter to `sweep_for_events` for a specific player to sweep for. This is used in door shuffle.

This should probably be its own PR prior to this one.

File: worlds/pokemon_rb/__init__.py [hunk 1]
```diff
@@ -209,196 +225,377 @@ def encode_name(name, t):
...
-                if (item.classification == ItemClassification.filler and self.multiworld.random.randint(1, 100)
-                        <= self.multiworld.trap_percentage[self.player].value and combined_traps != 0):
-                    item = self.create_item(self.select_trap())
-            if event:
-                self.multiworld.get_location(location.name, self.player).place_locked_item(item)
-            elif "Badge" not in item.name or self.multiworld.badgesanity[self.player].value:
-                item_pool.append(item)
+        self.dexsanity_table = [
+            *[True for _ in range(round(self.multiworld.dexsanity[self.player].value * 1.51))],
+            *[False for _ in range(151 - round(self.multiworld.dexsanity[self.player].value * 1.51))]
```
Comment (Berserker66) [inline worlds/pokemon_rb/__init__.py:231]:
```suggestion
            *(True for _ in range(round(self.multiworld.dexsanity[self.player].value * 1.51))),
            *(False for _ in range(151 - round(self.multiworld.dexsanity[self.player].value * 1.51)))
```

If they don't persist, they should be generators.

File: BaseClasses.py [hunk 2]
```diff
@@ -717,9 +717,10 @@ def can_reach(self,
                 spot = self.multiworld.get_region(spot, player)
         return spot.can_reach(self)
 
-    def sweep_for_events(self, key_only: bool = False, locations: Optional[Iterable[Location]] = None) -> None:
+    def sweep_for_events(self, key_only: bool = False, locations: Optional[Iterable[Location]] = None,
```
Comment (Berserker66) [inline BaseClasses.py:720]: I prefer my core changes in seperate PRs.

Comment (Alchav) [reply BaseClasses.py]: Removed

File: worlds/pokemon_rb/__init__.py [hunk 3]
```diff
@@ -209,196 +225,378 @@ def encode_name(name, t):
...
+
         if self.multiworld.old_man[self.player] == "early_parcel":
             self.multiworld.local_early_items[self.player]["Oak's Parcel"] = 1
             if self.multiworld.dexsanity[self.player]:
-                for location in [self.multiworld.get_location(f"Pokedex - {mon}", self.player)
-                                 for mon in poke_data.pokemon_data.keys()]:
-                    add_item_rule(location, lambda item: item.name != "Oak's Parcel" or item.player != self.player)
+                for mon in poke_data.pokemon_data:
+                    try:
+                        location = self.multiworld.get_location(f"Pokedex - {mon}", self.player)
```
Comment (Berserker66) [inline worlds/pokemon_rb/__init__.py:447]: multiworld.get_location causes a full recache of all locations in the entire multiworld on cache miss. Which mons exist should be tracked in a way that doesn't cause this to occur.

Comment (Berserker66) [reply worlds/pokemon_rb/__init__.py:447]: (There are several instances of this, I will not mark each in this review=

Comment (Alchav) [reply worlds/pokemon_rb/__init__.py:446]: Done

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**